platforms/allwinner-d1/core/src/ccu.rs

```diff
@@ -0,0 +1,260 @@
+//! This module provides a higher-level interface for the `Clock Controller Unit`.
+use d1_pac::CCU;
+use d1_pac::DMAC;
+use d1_pac::{SMHC0, SMHC1, SMHC2};
+use d1_pac::{SPI0, SPI_DBI};
+use d1_pac::{TWI0, TWI1, TWI2, TWI3};
+use d1_pac::{UART0, UART1, UART2, UART3, UART4, UART5};
+
+pub struct Ccu {
+    ccu: CCU,
+}
+
+/// Trait to be implemented for module clocks that can be gated and reset
+pub trait BusGatingResetRegister {
+    /// Enable or disable the clock reset bit
+    fn gating(ccu: &mut CCU, pass: bool);
+    /// Enable or disable the clock gating bit
+    fn reset(ccu: &mut CCU, deassert: bool);
+}
+
+macro_rules! set_module {
+    ($self: ident, $module: ident) => {
+        if ($self.ccu.$module.read().pll_en().bit_is_clear()) {
+            $self.ccu.$module.modify(|_, w| {
+                w.pll_ldo_en().enable();
+                w.pll_en().enable();
+                w
+            });
+
+            $self.ccu.$module.modify(|_, w| w.lock_enable().enable());
+
+            while $self.ccu.$module.read().lock().bit_is_clear() {
+                core::hint::spin_loop();
+            }
+            sdelay(20);
+
+            $self.ccu.$module.modify(|_, w| w.lock_enable().disable());
+        }
+    };
+}
+
+// TODO: should this move into the `Clint`?
+fn sdelay(delay_us: usize) {
+    let clint = unsafe { crate::clint::Clint::summon() };
+    let t = clint.get_mtime();
+    // TODO: confirm clock source of mtime
+    while clint.get_mtime() < (t + 24 * delay_us) {
+        core::hint::spin_loop()
+    }
+}
+
+impl Ccu {
+    pub fn new(ccu: CCU) -> Self {
+        Self { ccu }
+    }
+
+    pub fn release(self) -> CCU {
+        self.ccu
+    }
+
+    /// De-assert the reset bit and enable the clock gating bit for the given module
+    pub fn enable_module<MODULE: BusGatingResetRegister>(&mut self) {
+        MODULE::reset(&mut self.ccu, true);
+        sdelay(20);
+        MODULE::gating(&mut self.ccu, true);
+    }
+
+    /// Disable the clock gating bit and assert the reset bit for the given module
+    pub fn disable_module<MODULE: BusGatingResetRegister>(&mut self) {
+        MODULE::gating(&mut self.ccu, false);
+        // TODO: delay?
+        MODULE::reset(&mut self.ccu, false);
+    }
+
+    /// Allow modules to configure their own clock on a PAC level
+    // TODO: find a good abstraction so we don't need this anymore
+    pub fn borrow(&mut self) -> &mut CCU {
+        &mut self.ccu
+    }
+
+    /// Initialize the system clocks to the same default value that is also set by `xfel`
+    pub fn sys_clock_init(&mut self) {
+        // The clock initialization functions are ported to Rust, based on the C implementation in
+        // [xboot](https://github.com/xboot/xboot/blob/master/src/arch/riscv64/mach-d1/sys-clock.c)
+        self.set_pll_cpux_axi();
+        self.set_pll_periph0();
+        self.set_ahb();
+        self.set_apb();
```

Comment:
looking at the `xboot` source, i see that there's also a `set_dma` function, which is [also called in `sys_clock_init`](https://github.com/xboot/xboot/blob/6d6b93947b7fcb8c3924fedb0715c23877eedd5e/src/arch/riscv64/mach-d1/sys-clock.c#L209), and [resets + enables the DMA BGR](https://github.com/xboot/xboot/blob/6d6b93947b7fcb8c3924fedb0715c23877eedd5e/src/arch/riscv64/mach-d1/sys-clock.c#L156-L163). 

it looks like the reason we don't do that here is because the DMAC driver calls `enable()` on that register, here: https://github.com/tosc-rs/mnemos/blob/94199eb8946fc054aeb78a944818d7b9c73eeec5/platforms/allwinner-d1/core/src/dmac/mod.rs#L30

it might be worth including a comment here like
```suggestion
        self.set_apb();
        // This is where `xboot` resets and enables the DMA bus gating register.
        // We don't do that, because this is performed by the DMAC driver's 
        // initialization function, instead.
```
it might also be worth checking that we're actually setting the same bits they set when we call `enable<DMAC>()`?

Reply:
Indeed, I removed it here since we enable the peripheral clocks in the peripheral modules themselves.
The bits in `enable<DMAC>()` match those from `set_dma`.